packages/langchain/src/openai/util.ts

```diff
@@ -0,0 +1,205 @@
+        finish_reason: choice.finish_reason,
+        index: choice.index,
+        function_call: choice.message.function_call,
+        tool_calls: choice.message.tool_calls
+      }
+    })),
+    llmOutput: {
+      created: res.created,
+      id: res.id,
+      model: res.model,
+      object: res.object,
+      tokenUsage: {
+        completionTokens: res.usage.completion_tokens,
+        promptTokens: res.usage.prompt_tokens,
+        totalTokens: res.usage.total_tokens
+      }
+    }
+  };
+}
+
+/**
+ * Maps {@link BaseMessage} to OpenAI messages.
+ * @param message - The message to map.
+ * @returns The OpenAI chat Message.
+ */
+function mapBaseMessageToAzureOpenAiChatMessage(
+  message: BaseMessage
+): AzureOpenAiChatMessage {
+  // TODO: remove type casting, improve message.content handling
+  return removeUndefinedProperties<AzureOpenAiChatMessage>({
+    content: message.content,
+    name: message.name,
+    role: mapBaseMessageToRole(message),
+    function_call: message.additional_kwargs.function_call,
+    tool_calls: message.additional_kwargs.tool_calls,
+    tool_call_id:
+      message._getType() === 'tool'
+        ? (message as ToolMessage).tool_call_id
+        : undefined
+  } as AzureOpenAiChatMessage);
+}
+
+/**
+ * Checks if a given array is a structured tool array.
+ * @param tools - The array to check.
+ * @returns Whether the array is a structured tool array.
+ */
+function isStructuredToolArray(tools?: unknown[]): tools is StructuredTool[] {
+  return !!tools?.every(tool =>
+    Array.isArray((tool as StructuredTool).lc_namespace)
+  );
+}
+
+function mapToolChoice(
+  toolChoice?: LangChainToolChoice
+): ToolChoice | undefined {
+  if (toolChoice === 'auto' || toolChoice === 'none') {
+    return toolChoice;
+  }
+
+  if (typeof toolChoice === 'string') {
+    return {
+      type: 'function',
+      function: { name: toolChoice }
+    };
+  }
+}
+
+/**
+ * Maps LangChain's input interface to our own client's input interface
+ * @param client The LangChain OpenAI client
+ * @param options The LangChain call options
+ * @param messages The messages to be send
+ * @returns An AI SDK compatibile request
+ * @internal
+ */
+export function mapLangchainToAiClient(
+  client: AzureOpenAiChatClient,
+  options: AzureOpenAiChatCallOptions,
+  messages: BaseMessage[]
+): AzureOpenAiChatCompletionParameters {
+  return removeUndefinedProperties<AzureOpenAiChatCompletionParameters>({
+    messages: messages.map(mapBaseMessageToAzureOpenAiChatMessage),
+    max_tokens: client.max_tokens === -1 ? undefined : client.max_tokens,
+    temperature: client.temperature,
+    top_p: client.top_p,
+    logit_bias: client.logit_bias,
+    n: client.n,
+    stop: options?.stop ?? client.stop,
+    functions: isStructuredToolArray(options?.functions)
+      ? options?.functions.map(mapToolToOpenAiFunction)
+      : options?.functions,
+    tools: isStructuredToolArray(options?.tools)
+      ? options?.tools.map(mapToolToOpenAiTool)
```

Comment:
Step 3: So each entry in `options.tools` should have `schema` property. Check step 4.